Iptables idempotent

deargle opened the pull request:
Instead of a bash script, uses iptables cookbook to manage iptable rules. Makes it easier to let other recipes modify iptables.

Knockd refactored to use new iptables cookbook.

Ingreslock vuln added using new iptables cookbook.

jmartin-r7 commented on chef/cookbooks/iptables/CONTRIBUTING.md:
```diff
@@ -0,0 +1,2 @@
+Please refer to
```
remove file, that refers to instructions for other projects.

deargle replied:
I can remove it, but it came straight from a downloaded zip from https://supermarket.chef.io/cookbooks/iptables -- we should/could probably move to a git submodule approach, but how would that impact your testing workflows?

jmartin-r7 replied:
For testing adding submodule checkout is not a problem, it does add some complexity to for users.  If submodules are added, then `README.md` needs to be clear about usage and likely `build.sh` and `build.ps1` will need to be updated to check they have been pulled before attempting to execute.

deargle replied:
Eh, perhaps for a future PR. I removed CONTRIBUTING.md for now.

deargle replied:
Would you be okay installing the chefdk  in the testing run in order to use berks for dependency management? Where are the ubuntu test scripts -- or is it just the packer template and provisioning therein?

jmartin-r7 replied:
The only current tests for ubuntu are part of the chef recipe here https://github.com/rapid7/metasploitable3/tree/master/chef/cookbooks/metasploitable/test

Adding `chefdk` is likely a bit of overkill for the project at this time.  An additional note, I have found many users that choose to build these VMs themselves have had internet connectivity issues and adding more dependencies on infrastructure not provided by the OS and github is not a favored approach at this time.

deargle replied:
Hey here's an idea: we could move the vendor cookbooks to their own directory, and include that directory along with the primary metasploitable3 cookbooks. That's how I did it with my project's repo, see [here](https://github.com/deargle/gcp-vuln-server-behing-openvpn). Multiple cookbook paths can be passed to `chef-solo` for provisioning. Users would not have to install `chefdk` or use `berks` -- the vendor cookbooks could still be committed directly to the repo -- but it would be clearly delineated that those cookbooks were vendored. Would you be okay with github-related artifacts like CONTRIBUTING.md to remain in vendor cookbooks if they were in a directory clearly marked as such?

jmartin-r7 commented on chef/cookbooks/metasploitable/attributes/default.rb:
```diff
@@ -18,5 +18,6 @@
                                      :sinatra => 8181,
-                                     :samba => 445
+                                     :samba => 445,
+                                     :ssh => 22
```
I would suggest that `ssh` should not be added here.  The array here is `default[:metasploitable][:ports]` the intent of this variable is to be a list of ports that support vulnerable paths on the VM you might target with `metasploit-framework`.  Retaining the separation of `ssh` in the source is intended to make it clear the port is being allowed explicitly without a vuln attached to it.

deargle replied:
In that case, how do you feel about putting the port-opening rules contained within each vuln's recipe? In the name of modularization -- e.g., if the `readme_app` recipe isn't included, then its rule shouldn't be added to iptables, no?

jmartin-r7 replied:
I like the idea of extracting the rules into the recipes they are connected to.

deargle replied:
Done

deargle commented:
I removed `ingreslock`, and I moved all ip_rules and port-openings into the respective vuln recipes. the `metasploitable::iptables` recipe sets up three rules -- one for preserving established connections, one for ssh, and one for dropping everything. the `iptables` cookbook applies rules based on their alphabetical filename, with each rule getting an entry in `/etc/iptables.d/`, so I prefixed each iptables_rule with a number.

deargle commented:
Would it be helpful to you for merging this if I showed with `iptables -L` that the firewall rules are functionally equivalent to what they are on master rn?

jmartin-r7 commented:
Testing is in progress, this will land when complete.

deargle commented:
Bump, revisiting this since I'm about to start teaching class again. 

This should be relatively safe to merge as people are mostly exclusively having problems building the windows box, yeah? Also, merging this is prereq to getting ingreslock back in. 

and also bumping https://github.com/rapid7/metasploitable3/pull/424.

idempotence is important for development

deargle commented:
Woo! I'll redo an ingreslock pr now that this is in.

On Thu, Aug 20, 2020, 7:33 AM Jeffrey Martin <notifications@github.com>
wrote:

> Merged #423 <https://github.com/rapid7/metasploitable3/pull/423> into
> master.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rapid7/metasploitable3/pull/423#event-3675283289>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI6Y7PO5NXEOTDOMMUQWNTSBUQ2LANCNFSM4JGRDOBA>
> .
>
